File: scripts/native-metadata.ts
import { readFile } from "node:fs/promises";
import { resolve } from "node:path";
import {
  CLIENT_PROTOCOL_VERSION,
  SUPPORTED_PROTOCOL_MAX,
  SUPPORTED_PROTOCOL_MIN,
} from "../core/protocol-schemas/compatibility.generated.js";

export interface NativeBuildMetadata {
  schemaVersion: 1;
  /**
   * The version tag `release.yml` builds, without its `v`, from
   * `FROCKBOT_RELEASE`; null for any other build. It is the app's one version:
   * the Profile page shows it and the client's hello names it.
   */
  release: string | null;
  /**
   * What Flutter stamps into the native build: the release's
   * `major.minor.patch`, or `pubspec.yaml`'s placeholder outside a release.
   */
  app: {
    versionName: string;
    buildNumber: number;
    version: string;
  };
  hostedOrigin: string;
  clientProtocol: number;
  compatibility: {
    protocolMin: number;
    protocolMax: number;
  };
}

interface CompatibilitySource {
  clientProtocol: unknown;
  protocolMin: unknown;
  protocolMax: unknown;
}

const defaultCompatibility: CompatibilitySource = {
  clientProtocol: CLIENT_PROTOCOL_VERSION,
  protocolMin: SUPPORTED_PROTOCOL_MIN,
  protocolMax: SUPPORTED_PROTOCOL_MAX,
};

function positiveProtocol(value: unknown, label: string): number {
  if (
    !Number.isInteger(value) ||
    (value as number) < 1 ||
    (value as number) > 65_535
  ) {
    throw new Error(`${label} must be an integer from 1 to 65535.`);
  }
  return value as number;
}

function compatibilityMetadata(source: CompatibilitySource) {
  const protocolMin = positiveProtocol(source.protocolMin, "protocolMin");
  const protocolMax = positiveProtocol(source.protocolMax, "protocolMax");
  const clientProtocol = positiveProtocol(
    source.clientProtocol,
    "clientProtocol",
  );
  if (protocolMin > protocolMax) {
    throw new Error("protocolMin must not exceed protocolMax.");
  }
  if (clientProtocol < protocolMin || clientProtocol > protocolMax) {
    throw new Error(
      `Client protocol ${clientProtocol} is outside the supported ${protocolMin}..${protocolMax} range.`,
    );
  }
  return { clientProtocol, protocolMin, protocolMax };
}

/** The pattern `ClientHello.nativeVersion` accepts, so a release can always say its name. */
const RELEASE =
  /^(0|[1-9]\d{0,5})\.(0|[1-9]\d{0,5})\.(0|[1-9]\d{0,5})(-[0-9A-Za-z.-]{1,20})?$/u;

function releaseVersion(value: string | undefined) {
  if (value === undefined || value === "") return null;
  const match = RELEASE.exec(value);
  if (!match || value.length > 32) {
    throw new Error(
      `FROCKBOT_RELEASE must be a version tag without its \`v\`, such as 0.7.163 or 0.8.0-rc.1, not ${JSON.stringify(value)}.`,
    );
  }
  return { release: value, versionName: `${match[1]}.${match[2]}.${match[3]}` };
}

function appIdentity(document: unknown) {
  if (!document || typeof document !== "object") {
    throw new Error("apps/native/pubspec.yaml must contain a YAML mapping.");
  }
  const version = (document as { version?: unknown }).version;
  if (typeof version !== "string") {
    throw new Error("apps/native/pubspec.yaml must define a string version.");
  }
  const match = /^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)\+([1-9]\d*)$/u.exec(
    version,
  );
  if (!match) {
    throw new Error(
      "apps/native/pubspec.yaml version must be `<major>.<minor>.<patch>+<positive build>`.",
    );
  }
  const versionName = `${match[1]}.${match[2]}.${match[3]}`;
  const buildNumber = Number(match[4]);
  if (!Number.isSafeInteger(buildNumber)) {
    throw new Error("apps/native/pubspec.yaml build number is too large.");
  }
  return { versionName, buildNumber, version };
}

function hostedOrigin(document: unknown): string {
  const hostnames = (
    document as {
      workers?: { app?: { hostnames?: unknown } };
    } | null
  )?.workers?.app?.hostnames;
  if (
    !Array.isArray(hostnames) ||
    typeof hostnames[0] !== "string" ||
    hostnames[0].length === 0
  ) {
    throw new Error(
      "deployments/hosted.json must name a first application hostname.",
    );
  }
  const origin = new URL(`https://${hostnames[0]}`);
  if (
    origin.hostname !== hostnames[0] ||
    origin.username ||
    origin.password ||
    origin.port ||
    origin.pathname !== "/"
  ) {
    throw new Error("deployments/hosted.json application hostname is invalid.");
  }
  return origin.origin;
}

export async function readNativeMetadata(
  repositoryRoot = resolve(import.meta.dirname, ".."),
  compatibilitySource: CompatibilitySource = defaultCompatibility,
  releaseSource = process.env.FROCKBOT_RELEASE,
): Promise<NativeBuildMetadata> {
  const [pubspecSource, hostedSource] = await Promise.all([
    readFile(resolve(repositoryRoot, "apps/native/pubspec.yaml"), "utf8"),
    readFile(resolve(repositoryRoot, "deployments/hosted.json"), "utf8"),
  ]);
  const placeholder = appIdentity(Bun.YAML.parse(pubspecSource));
  const compatibility = compatibilityMetadata(compatibilitySource);
  const tagged = releaseVersion(releaseSource);
  const versionName = tagged?.versionName ?? placeholder.versionName;
  return {
    schemaVersion: 1,
    release: tagged?.release ?? null,
    app: {
      versionName,
      buildNumber: placeholder.buildNumber,
      version: `${versionName}+${placeholder.buildNumber}`,
    },
    hostedOrigin: hostedOrigin(JSON.parse(hostedSource) as unknown),
    clientProtocol: compatibility.clientProtocol,
    compatibility: {
      protocolMin: compatibility.protocolMin,
      protocolMax: compatibility.protocolMax,
    },
  };
}

if (import.meta.main) {
  const rootIndex = process.argv.indexOf("--root");
  if (rootIndex >= 0 && !process.argv[rootIndex + 1]) {
    throw new Error("--root requires a repository path.");
  }
  const repositoryRoot =
    rootIndex >= 0 ? resolve(process.argv[rootIndex + 1]!) : undefined;
  process.stdout.write(
    `${JSON.stringify(await readNativeMetadata(repositoryRoot), null, 2)}\n`,
  );
}

File: scripts/native_metadata.py
"""Checked access to the native build metadata command.

Python release tools consume this decoder instead of scraping Dart, YAML, or
generated TypeScript. The Bun command owns source parsing and compatibility
checks; this boundary rejects incomplete or mistyped command output.
"""

from dataclasses import dataclass
import json
from pathlib import Path
import subprocess
from typing import Optional


@dataclass(frozen=True)
class NativeMetadata:
    release: Optional[str]
    version_name: str
    build_number: int
    version: str
    hosted_origin: str
    client_protocol: int
    protocol_min: int
    protocol_max: int


def _integer(value, label):
    if type(value) is not int:
        raise RuntimeError(f"Native metadata {label} must be an integer.")
    return value


def decode_native_metadata(source):
    try:
        document = json.loads(source)
        app = document["app"]
        compatibility = document["compatibility"]
        metadata = NativeMetadata(
            release=document["release"],
            version_name=app["versionName"],
            build_number=_integer(app["buildNumber"], "app.buildNumber"),
            version=app["version"],
            hosted_origin=document["hostedOrigin"],
            client_protocol=_integer(document["clientProtocol"], "clientProtocol"),
            protocol_min=_integer(compatibility["protocolMin"], "compatibility.protocolMin"),
            protocol_max=_integer(compatibility["protocolMax"], "compatibility.protocolMax"),
        )
    except (json.JSONDecodeError, KeyError, TypeError) as failure:
        raise RuntimeError("Native metadata command returned an invalid document.") from failure
    if document.get("schemaVersion") != 1:
        raise RuntimeError("Native metadata command returned an unsupported schema version.")
    for label, value in (
        ("app.versionName", metadata.version_name),
        ("app.version", metadata.version),
        ("hostedOrigin", metadata.hosted_origin),
    ):
        if not isinstance(value, str) or not value:
            raise RuntimeError(f"Native metadata {label} must be a non-empty string.")
    if metadata.release is not None and (not isinstance(metadata.release, str) or not metadata.release):
        raise RuntimeError("Native metadata release must be a non-empty string or null.")
    if metadata.version != f"{metadata.version_name}+{metadata.build_number}":
        raise RuntimeError("Native metadata app version does not match its name and build number.")
    if not metadata.hosted_origin.startswith("https://"):
        raise RuntimeError("Native metadata hostedOrigin must be HTTPS.")
    if not metadata.protocol_min <= metadata.client_protocol <= metadata.protocol_max:
        raise RuntimeError("Native metadata client protocol is outside its compatibility range.")
    return metadata


def read_native_metadata(repository_root):
    root = Path(repository_root).resolve()
    command = Path(__file__).with_name("native-metadata.ts")
    result = subprocess.run(
        ["bun", command, "--root", root],
        cwd=root,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        detail = result.stderr.strip() or result.stdout.strip() or f"exit {result.returncode}"
        raise RuntimeError(f"Could not read native build metadata: {detail}")
    return decode_native_metadata(result.stdout)
